File: python/breadth-first-search/SurroundedRegions.py
from typing import List
from collections import deque

class Solution:
    def solve(self, board: List[List[str]]) -> None:

        rows, cols = len(board), len(board[0])
        distances = [(-1, 0), (0, -1), (0, 1), (1, 0)]

        def is_valid_cell(row, col):
            return 0 <= row < rows and 0 <= col < cols
        
        queue = deque()

        for i in range(rows):
            for j in range(cols):
                if i == 0 or i == rows - 1 or j == 0 or j == cols - 1:
                    if board[i][j] == 'O':
                        queue.append((i, j))
                        board[i][j] = 'V'
                    
        while queue:
            row, col = queue.popleft()

            for direction in distances:
                new_row, new_col = row + direction[0], col + direction[1]
                if is_valid_cell(new_row, new_col) and board[new_row][new_col] == 1:
                    queue.append((new_row, new_col))
                    board[new_row][new_col] = 'V'

        
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == 'V':
                    board[i][j] = 'O'
                else:
                    board[i][j] = 'X'
    

File: cpp/breadth-first-search/SurroundedRegions.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int row, col;
    Node(int row, int col) {
        this->row = row;
        this->col = col;
    }
};

class Solution {
public:

    vector<vector<int>> directions = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};

    bool isValidCell(int row, int col, int rows, int cols) {
        if(row >= 0 && row < rows && col >= 0 && col < cols) 
            return true;
        else 
            return false;
    }

    void solve(vector<vector<char>>& board) {

        int rows = board.size();
        int cols = board[0].size();
        
        queue<Node> queue;

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(i == 0 || i == rows - 1 || j == 0 || j == cols - 1) {
                    if(board[i][j] == 'O') {
                        queue.push(Node(i, j));
                        board[i][j] = 'V';
                    }
                }
            }
        }

        while(!queue.empty()) {
            Node node = queue.front();
            queue.pop();

            for(auto& direction : directions) {
                int newRow = node.row + direction[0];
                int newCol = node.col + direction[1];
                if(isValidCell(newRow, newCol, rows, cols) && board[newRow][newCol] == 'O') {
                    queue.push(Node(newRow, newCol));
                    board[newRow][newCol] = 'V';
                }
            }
        }

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(board[i][j] == 'V') {
                    board[i][j] = 'O';
                } else {
                    board[i][j] = 'X';
                }
            }
        }
        
    }
};